### PR TITLE
Add `password_cmd` feature

### PR DESCRIPTION
**Pull request by bcmyers**
This PR adds the ability for the user to add a `password_cmd` entry to his/her config representing a bash command that `spotifyd` will call upon launch in order to obtain the user's password. 

It's a valuable feature because it makes it so the user doesn't have to store his/her password in the config file. In addition, unlike the dbus/keyring feature, it works with [pass](https://www.passwordstore.org/), which is the password management utility I use and which is, I believe, fairly popular on linux and macOS.

As a side-effect of thinking about subprocesses, I also improved the code concerning the `onevent` feature. Any event-driven subprocess now, when exiting unsuccessfully, will log the contents of it's stderr output as well as the original command that was called. 

In addition, I think the way subprocess commands were being "parsed" before would break under certain circumstances. You were assuming you could take the user's input from `onevent` and just call `split_whitespace` on it to get the args to pass to `Command`. I don't think that'll work if the user's command has a bunch of nested quotes and/or double-quotes; so I changed the way we're launching subprocesses to ```Command::new("bash").arg("-c").arg(user_input_str)...```.

If you have any questions, of course don't hesitate to reach out. Looking forward to seeing if this might be something you'd like to include.

Thanks!

### CODE REVIEW

**Comment by SimonPersson** [comment]
Thanks for this! Looks nice, I left only one minor stylistic comment.

> In addition, I think the way subprocess commands were being "parsed" before would break under certain circumstances. You were assuming you could take the user's input from onevent and just call split_whitespace on it to get the args to pass to Command. I don't think that'll work if the user's command has a bunch of nested quotes and/or double-quotes

You are right, that was pretty silly of me :) I don't like having bash hard coded like that though. That will definitely be confusing to users of any other shell. My preferred solution would be to just change the documentation for `onevent`, to make it clear that we are not going to parse the command, and it will not be passed to a shell. Users can specify the path to a shell script in `onevent`, if the shell is desired.

**Comment by bcmyers** [comment]
That makes sense. We shouldn't hard code bash. That said, I think we can still get the shell to do the command parsing for us so that user is able to write in their config file or in a CLI option exactly what they would write in their shell and it just works. 

We just need to figure out which shell the user would like to use, which we can do with something like the below.

What do you think? I tested this on linux, FreeBSD and macOS and it appears to work. 

And clearly we would also need to document for the user what this is doing, i.e. we first look for the shell specified in the SHELL environment variable and, if that's not set, we default to the system's default shell for the user.

```rust
#![cfg(unix)]

use std::cell::RefCell;
use std::env;
use std::ffi::CStr;

extern "C" {
    fn getlogin_r(buf: *mut libc::c_char, size: libc::size_t) -> libc::c_int;
}

thread_local! {
    static BUF_USERNAME: RefCell<[libc::c_char; 255]> = RefCell::new([0; 255]);
}

pub fn get_shell<S>(username: S) -> Option<String>
where
    S: AsRef<str>,
{
    // First look for the user's preferred shell using the SHELL environment variable...
    if let Ok(shell) = env::var("SHELL") {
        log::trace!("Found shell {:?} using SHELL environment variable.", shell);
        return Some(shell);
    }

    // If the SHELL environment variable is not set and we're on linux or one of the BSDs,
    // try to obtain the default shell from `/etc/passwd`...
    #[cfg(not(target_os = "macos"))]
    {
        use std::fs::File;
        use std::io::{self, BufRead};

        let file = File::open("/etc/passwd").ok()?;
        let reader = io::BufReader::new(file);
        // Each line of `/etc/passwd` describes a single user and contains seven colon-separated fields:
        // "name:password:UID:GID:GECOS:directory:shell"
        for line in reader.lines() {
            let line = line.ok()?;
            let mut iter = line.split(":");
            if let Some(user) = iter.nth(0) {
                if user == username.as_ref() {
                    let shell = iter.nth(5)?;
                    log::trace!("Found shell {:?} using /etc/passwd.", shell);
                    return Some(shell.into());
                }
            }
        }
    }

    // If the SHELL environment variable is not set and on we're on macOS,
    // query the Directory Service command line utility (dscl) for the user's shell,
    // as macOS does not use the /etc/passwd file...
    #[cfg(target_os = "macos")]
    {
        use std::process::Command;

        let output = Command::new("dscl")
            .args(&[".", "-read", &format!("/Users/{}", username), "UserShell"])
            .output()
            .ok()?;
        if output.status.success() {
            let stdout = std::str::from_utf8(&output.stdout).ok()?;
            // The output of this dscl command should be:
            // "UserShell: /path/to/shell"
            if stdout.starts_with("UserShell: ") {
                let shell = stdout.split_whitespace().nth(1)?;
                log::trace!("Found shell {:?} using dscl command.", shell);
                return Some(shell.to_string());
            }
        }
    }

    None
}

pub fn get_username() -> Option<String> {
    BUF_USERNAME.with(|refcell| {
        let mut buf = refcell.borrow_mut();
        let ret = unsafe { getlogin_r(buf.as_mut_ptr() as _, buf.len() as _) };
        if ret != 0 {
            return None;
        }
        let cstr = unsafe { CStr::from_ptr(buf.as_ptr()) };
        let username = cstr.to_string_lossy().to_string();
        Some(username)
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        env_logger::init();

        let username = get_username().unwrap();
        log::info!("Found username: {:?}", username);

        let shell = get_shell(username).unwrap();
        log::info!("Found preferred shell: {:?}", shell);
    }
}
```

**Comment by bcmyers** [comment]
I'll add a commit that incorporates this so you can look at it in the context of the whole.

**Comment by SimonPersson** [comment]
Would using `sh -c` instead of `bash -c` be an acceptable solution to you? It might not be the users preferred shell, but it will at least be there. To use shell-specific features you could specify that shell in `onevent`.

**Comment by bcmyers** [comment]
If you don't like the latest commit (which is totally fine by me; I don't have a super strong preferences here), then, sure, let's just use `sh`. That said, I would marginally prefer to do what this latest commit does, i.e.:

* Use the shell indicated by the `SHELL` environment variable, if present.
* If the `SHELL` environment variable is not present, use the user's default shell.
* Only if all of that fails, which I don't think is even possible, then actually hard-code `sh`.

**Comment by SimonPersson** [comment]
All right, let's try it this way. It's a bit more code for a relatively minor feature, but we can always revert to just `sh` if it's causing us issues.

File: src/config.rs [hunk 1]
```diff
@@ -138,11 +140,14 @@ pub fn get_config<P: AsRef<Path>>(config_path: Option<P>, matches: &Matches) ->
...
-        matches.opt_str(field).or_else(|| {
-            spotifyd
-                .and_then(|s| s.get(field).map(String::clone))
-                .or_else(|| global.and_then(|s| s.get(field).map(String::clone)))
-        })
+        if matches.opt_defined(field) {
+            if let Some(val) = matches.opt_str(field) {
+                return Some(val);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (bcmyers) [reply src/config.rs]: I'll switch the style back to using `or_else` combinators, but the call to `opt_defined` is necessary as the code is now. 

`opt_str` panics if the field your asking for is not a CLI option and this code is being used to query `password_cmd`, which is, as it's currently written, not a CLI option (you can only specify it in the config file). 

That brings up a good question, though. Would you like `password_cmd` to be a CLI option as well? `onevent`, which is similar, is both a CLI option and a config file option; so there's an argument to make `password_cmd` a CLI option just for the coherence / symmetry.

Let me know what you think. In the meantime, I'll change this back to using combinators.

**Comment by bcmyers** [comment]
Speaking of CLI options, have you thought about switching from `getopts` to [`structopt`](https://crates.io/crates/structopt), which seems to have become the default standard in the Rust ecosystem. Doing that switch feel like it belongs in a separate PR, but I'd be happy to tackle that as well, if you're interested.

**Comment by SirWindfield** [comment]
@bcmyers I actually started working on this already. But I do not have that much time now since I just started with my finals thesis.
I would be glad if you could take that part over!

The goal would be to have one strucr that combines the config and the cli options. You can make it derive from serde as well as structopt.

As of which one should have priority, I would say cli options go over config file options. You could read the config file at runtime and write a function that just merges them, taking values from the cli if they are available.

I got some more ideas for future work on this refactoring, although they aren't high priority right now.

**Comment by SimonPersson** [comment]
Thanks a lot for this @bcmyers. I'll merge this now.

As @SirWindfield commented, we have discussed refactoring the config system before (https://github.com/Spotifyd/spotifyd/issues/233, https://github.com/Spotifyd/spotifyd/issues/262#issuecomment-507586308), and reached the conclusion that we would like unify the configuration file options with the command line options. This could involve moving to `structopt`. Feel free to work on that if you have time! :)

**Comment by bcmyers** [comment]
👍

**Comment by bcmyers** [comment]
@SirWindfield - I haven't forgotten about trying to sort out the config / cli situation. One question I have as I'm going over the code, however ... What is the purpose of the Discovery Stream? What is this used for? Some sort of alternative authentication mechanism?

**Comment by bcmyers** [comment]
Pinging @SimonPersson as well, as perhaps you can help answer my question above.

**Comment by SirWindfield** [comment]
I'm throwing some guesses around but as it is part of `librespot` I would say that it is responsible for announcing itself as a streamable device for spotify. But I guess @SimonPersson has better knowledge about that :smile:

P.S.: Is there a nicer way of implementing `dbus` in Rust? The current file is, in all honesty, quite overwhelming :smile:. Do you have any idea why skip and previous don't work? Apparently you are calling the right API methods... :thinking:

**Pull request merged**